### PR TITLE
Fixes numerous validating and defaulting. Improve tests

### LINKED ISSUE #92: Defaulting is incomplete and validation seems to miss some fields.
**Description**

If you leave out keyless.url it is not defaulted nor validated. ctlog for keyless is probably the same as well. So, this issue tracks complete validation for all the fields.

**Version**
head

### PR DESCRIPTION
**Pull request by vaikas**
#### Summary
* Fix #92 
* Clean up some over-verbose validation error messages
* [tester] Add a check that the image matches CIP when using the tester CLI
* [tester] Dump out the CIP that's being used to see defaulted fields
* Validate errors for invalid CIP is what we expect them to be. Introduce a way to specify expected errors in testdata CIPs

#### Release Note
* Fix #92 and other validation / defaulting problems
* Clean up some over-verbose validation error messages
* [tester] Add a check that the image matches CIP when using the tester CLI
* [tester] Dump out the CIP that's being used to see defaulted fields
* Validate errors for invalid CIP is what we expect them to be. Introduce a way to specify expected errors in testdata CIPs
*

### CODE REVIEW

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go [hunk 1]
```diff
@@ -154,19 +160,15 @@ func (key *KeyRef) Validate(ctx context.Context) *apis.FieldError {
...
+	if keyless.URL == nil && keyless.CACert == nil {
+		errs = errs.Also(apis.ErrMissingOneOf("url", "ca-cert"))
 	}
 
 	// TODO: Are these really mutually exclusive?
 	if keyless.URL != nil && keyless.CACert != nil {
 		errs = errs.Also(apis.ErrMultipleOneOf("url", "ca-cert"))
 	}
 
-	if keyless.Identities != nil && len(keyless.Identities) == 0 {
```
Comment (hectorj2f) [inline pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:166]: Question: won't we need to handle this condition ?

**Review by hectorj2f — APPROVED** [review]

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go (hunk 1 above)
Comment (vaikas) [reply pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:166]: It's marked as optional and it's fine to say that you don't care who signed (or attested something) :) If you don't give these, then as long as somebody signed or attested will be accepted :)

Comment (hectorj2f) [reply pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:166]: Gotcha! 👍🏻

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_defaults.go [hunk 2]
```diff
@@ -29,5 +31,8 @@ func (spec *ClusterImagePolicySpec) SetDefaults(ctx context.Context) {
 		if authority.Name == "" {
 			spec.Authorities[i].Name = fmt.Sprintf("authority-%d", i)
 		}
+		if authority.Key == nil && authority.Static == nil && authority.Keyless != nil && authority.Keyless.CACert == nil && authority.Keyless.URL == nil {
+			authority.Keyless.URL = apis.HTTPS("fulcio.sigstore.dev")
+		}
```
Comment (JAORMX) [inline pkg/apis/policy/v1alpha1/clusterimagepolicy_defaults.go:36]: does the ctlog url need to be set as well? we probably need to point to `rekor.sigstore.dev`

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go [hunk 3]
```diff
@@ -154,19 +160,15 @@ func (key *KeyRef) Validate(ctx context.Context) *apis.FieldError {
 
 func (keyless *KeylessRef) Validate(ctx context.Context) *apis.FieldError {
 	var errs *apis.FieldError
-	if keyless.URL == nil && keyless.Identities == nil && keyless.CACert == nil {
-		errs = errs.Also(apis.ErrMissingOneOf("url", "identities", "ca-cert"))
+	if keyless.URL == nil && keyless.CACert == nil {
+		errs = errs.Also(apis.ErrMissingOneOf("url", "ca-cert"))
```
Comment (JAORMX) [inline pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:164]: Why is the validation for `identities` being removed?

File: pkg/apis/policy/v1beta1/clusterimagepolicy_defaults.go [hunk 4]
```diff
@@ -29,5 +31,8 @@ func (spec *ClusterImagePolicySpec) SetDefaults(ctx context.Context) {
 		if authority.Name == "" {
 			spec.Authorities[i].Name = fmt.Sprintf("authority-%d", i)
 		}
+		if authority.Key == nil && authority.Static == nil && authority.Keyless != nil && authority.Keyless.CACert == nil && authority.Keyless.URL == nil {
+			authority.Keyless.URL = apis.HTTPS("fulcio.sigstore.dev")
+		}
```
Comment (JAORMX) [inline pkg/apis/policy/v1beta1/clusterimagepolicy_defaults.go:36]: does the ctlog url need to be set as well? we probably need to point to `rekor.sigstore.dev`

**Review by JAORMX — CHANGES_REQUESTED** [review]

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go (hunk 3 above)
Comment (vaikas) [reply pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:164]: It's not a required field. Just like with cosign the CLI if you do not specify `--certificate-email` or `--certificate-oidc-issuer` the signature / attestation is verified, but it's not checked who signed it.
Identities are still validated if they are specified.

**Review by jdolitsky — APPROVED** [review]
👍

File: pkg/apis/policy/v1beta1/clusterimagepolicy_defaults.go (hunk 4 above)
Comment (JAORMX) [reply pkg/apis/policy/v1beta1/clusterimagepolicy_defaults.go:36]: Sounds good to me

**Review by JAORMX — APPROVED** [review]

**Comment by JAORMX** [comment]
/lgtm

Thanks, the comments make sense and this can merge

**Pull request merged**